components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts

```diff
@@ -0,0 +1,311 @@
+      per_page: this.per_page,
+      server_id: this.serverId,
+      org_id: this.orgId
+    };
+    this.loadCookbookConstraint();
+
+    this.store.select(saveStatus)
+    .pipe(
+      takeUntil(this.isDestroyed),
+      filter(state => state === EntityStatus.loadingSuccess))
+      .subscribe(state => {
+        this.creating = false;
+        if (state === EntityStatus.loadingSuccess) {
+          this.store.dispatch(new GetEnvironments(payload));
+          this.closeCreateModal();
+        }
+      });
+
+    combineLatest([
+      this.store.select(saveStatus),
+      this.store.select(saveError)
+    ]).pipe(
+      takeUntil(this.isDestroyed),
+      filter(([state, error]) => state === EntityStatus.loadingFailure && !isNil(error)))
+      .subscribe(([_, error]) => {
+        if (error.status === HttpStatus.CONFLICT) {
+          this.conflictErrorEvent.emit(true);
+          this.conflictError = true;
+        } else {
+          this.store.dispatch(new GetEnvironments(payload));
+          this.creating = false;
+
+          // Close the modal on any other error because it will be displayed in the banner.
+          this.closeCreateModal();
+        }
+        this.creating = false;
+      });
+  }
+
+  ngOnDestroy(): void {
+    this.isDestroyed.next(true);
+    this.isDestroyed.complete();
+  }
+
+  closeCreateModal(): void {
+    this.resetCreateModal();
+    this.visible = false;
+  }
+
+  createEnvironment() {
+    this.creating = true;
+    const environment: CreateEnvironmentPayload = {
+      org_id: this.orgId,
+      server_id: this.serverId,
+      name: this.detailsFormGroup.controls['name'].value,
+      description: this.detailsFormGroup.controls['description'].value,
+      cookbook_versions: this.constraintArray.length ? this.toDisplay(this.constraintArray) : {},
+      default_attributes: JSON.parse(this.defaultAttrFormGroup.controls['default'].value),
+      override_attributes: JSON.parse(this.overrideAttrFormGroup.controls['override'].value)
+    };
+
+    this.store.dispatch(
+      new CreateEnvironment({environment: environment})
+    );
+  }
+
+  handleInput(event: { target: { value: string } } ): void {
+    this.nameExist = this.environmentsList.some(el => el.name === event.target.value);
+  }
+
+  onChangeDefaultJson(event: { target: { value: string } } ) {
+    // get value from text area
+    const newValue = event.target.value;
+    try {
+      // parse it to json
+      JSON.parse(newValue);
+      this.defaultAttrParseError = false;
+    } catch (ex) {
+      // set parse error if it fails
+      this.defaultAttrParseError = true;
+    }
+  }
+
+  onChangeOverrideJson(event: { target: { value: string } } ) {
+    // get value from text area
+    const newValue = event.target.value;
+    try {
+      // parse it to json
+      JSON.parse(newValue);
+      this.overrideAttrParseError = false;
+    } catch (ex) {
+      // set parse error if it fails
+      this.overrideAttrParseError = true;
```

Comment:
Comments should only be used for clarification or explication--these are not providing any info to the reader, so are not needed here. I wrote a possibly helpful article on best practices of writing comments: [Fighting Evil in Your Code: Comments on Comments](https://www.red-gate.com/simple-talk/opinion/opinion-pieces/fighting-evil-code-comments-comments/)

Reply:
removed these comments